Improve the doc for volume actual size and space usage (#487)
longhorn/website: content/docs/1.2.4/volumes-and-nodes/volume-size.md

    - For the overlapping part during the coalesce, the newer data (data#2) will be retained in the blocks. Then some historical data is removed and the volume gets shrunk (from 6.1Gi to 4.65Gi in the example).
{{< figure src="/img/screenshots/volumes-and-nodes/volume-size-example-4.png" >}}

## Space Configuration Suggestions for Volumes

1. Expand disks in the nodes so that there is an enough buffer for Longhorn volumes.
    - Each snapshot or volume head contains at most `size` space. A quick esitimation for maximum space consumption estimation of a volume is
  
        ```
        maximum space consumption = (N + 1) x volume nominal size
        ```

        where N is the number of snapshots retained, and 1 is for the system snapshot.

2. Do not retain too many snapshots for the volumes.

3. Clean up snapshots will help reclaim disk space. There are two ways to clean up snapshots:
    - Delete the snapshots manually via Longhorn UI.
    - Set a snapshot recurring job with retention 1, then the snapshots will be cleaned up automatically.
 
    Also, notice that the extra space, up to volume nominal `size`, is required during snapshot cleanup and merge.

> 4. Set `Storage Over Provisioning Percentage` to be smaller than 100% if there are write-heavy volumes.

Reviewer: Maybe you can combine 1st item with 4th item.
- Reserve enough free space in disks as buffers in case of the actual size of existing volumes keep growing up. 
  - A quick estimation for maximum space consumption estimation of a volume is: 
    ```
    (N + 1 + 1) x head/snapshot average actual size
    ```
    where N is the number of snapshots retained, and the 1st 1 is for the system snapshot, and 2nd 1 is for the tmp space may required by snapshot deletion. 
  - If there are heavy writing tasks for volumes, `head/snapshot average actual size` would be volume the nominal size. And in this case, it's better to set `Storage Over Provisioning Percentage` to be smaller than 100%
to avoid disk space exhaustion.

4. Then, rewrite the 50Gi data, and the `df` command in the filesystem shows 50Gi used space again. Howerver, the `actual size` is increased by 50Gi and becomes 100Gi.

    - After deletion, filesystem may or maynot reuse the recently freed blocks from recent deleted files according to the filesystem design and please refer to [Block allocation strategies of various filesystems](https://www.ogris.de/blkalloc). If the volume nominal `size` is 100Gi, the `actual size` in the end would range from 50Gi to 100Gi since the filesystem may or maynot reuse the freed blocks. On the other hand, if the volume nominal `size` is 60Gi, the `actual size` at the end would range from 50Gi to 60Gi, because the filesystem has to reuse the freed blocks in the 2nd round of writing. Thus, allocating an appropriate nominal `size` for a volume that holds heavy writing tasks according to the IO pattern would make disk space usage more efficient.

> ### An Example That Helps Understand Volume `Size` and `Actual Size`:

Reviewer: There are overlapping parts between the 2 examples. I think we can combine them and emphasize the following key points (with more illustrations):
1. Dirty block not releasing issue: Longhorn does not support TRIM/UNMAP operations. Hence deleting files from filesystems won't lead to volume actual size decreasing/shrinking. 
2. Dirty block not reusing issue: Deleting then writing new files would lead to the actual size keeps increasing. Since the filesystems may not reuse the dirty blocks (the recently freed blocks from recent deleted files). Thus, allocating an appropriate nominal `size` for a volume that holds heavy writing tasks according to the IO pattern would make disk space usage more efficient. 
3. Deleting snapshots **may** (not guarantee) eliminate the overlapping part of the used blocks (no matter the blocks are recently released blocks by fs or still hold historical data).

    - The `actual size` is still 2Gi.
{{< figure src="/img/screenshots/volumes-and-nodes/volume-size-example-2.png" >}}

4. Write 4Gi data (data#2) in the volume mount, then take one more snapshot (snapshot#2). See Figure 3 of the illustration.
    - Now the `actual size` is 6Gi, which is greater than the nominal `size`.
    - There is overlapping between 2 snapshots in block level (See the 2 snapshots in Figure 3), since the data#1 is marked as removed in snapshot#2 then the filesystem reuses the space.

{{< figure src="/img/screenshots/volumes-and-nodes/volume-size-example-3.png" >}}
5. Delete snapshot#1 and wait for snapshot purge complete. See Figure 4 of the illustration.
    - Here Longhorn actually coalesces the snapshot#1 with the snapshot#2.
    - For the overlapping part during the coalesce, the newer data (data#2) will be retained in the blocks. Then some historical data is removed and the volume gets shrunk (from 6.1Gi to 4.65Gi in the example).
{{< figure src="/img/screenshots/volumes-and-nodes/volume-size-example-4.png" >}}

> ## Space Configuration Suggestions for Volumes

Reviewer: Add suggestion:
4. An appropriate the spec/nominal size

- Since all historical data stored in the snapshots and active data will be calculated into the actual size, the final value can be greater than the nominal size.
- The actual size will be shown only when the volume is running.

## Example

In the example, we will explain how volume `size` and `actual size` get changed after a bunch of IO and snapshot related operations.

> The illustration presents the file organization of **one replica**. The volume head and snapshots are actually sparse files, which we mentioned above.

> {{< figure src="/img/screenshots/volumes-and-nodes/volume-size-illustration.png" >}}

Reviewer: NIT: You can replace the description of the gray part with `the blocked freed by the file deletion of filesystem but still considered as allocated in Longhorn's perspective`

    > The volume head holds the latest data of the volume only, while each snapshot may store historical data as well as active data, which consumes at most size space. Therefore, the volume `actual size`, which is the size sum of the volume head and all snapshots, is possibly bigger than the size specified by users.
    >
    > Even if users will not take snapshots for volumes, there are operations like rebuilding, expansion, or backing up that would lead to system (hidden) snapshot creation. As a result, volume `actual size` being larger than size is unavoidable under some use cases.

{{< figure src="/img/screenshots/volumes-and-nodes/volume-size-illustration-fig5.png" >}}

8. Delete snapshot#1 and wait for snapshot purge complete. See Figure 7 of the illustration.
    - Here Longhorn actually coalesces the snapshot#1 with the snapshot#2.
    - For the overlapping part during the coalesce, the newer data (data#2) will be retained in the blocks. Then some historical data is removed and the volume gets shrunk (from 16.3 Gi to 11.4 Gi in the example).

> {{< figure src="/img/screenshots/volumes-and-nodes/volume-size-illustration-fig6.png" >}}

Reviewer: NIT: We can add a brief summary at the end, which includes the 3 key points I mentioned last time.

     >
     > Thus, allocating an appropriate nominal `size` for a volume that holds heavy writing tasks according to the IO pattern would make disk space usage more efficient.

{{< figure src="/img/screenshots/volumes-and-nodes/volume-size-illustration-fig3.png" >}}

5. Take a snapshot (snapshot#1). See Figure 4 of the illustration.
    - Now data#1 is stored in snapshot#1.
    - The new volume head size is almost 0.
    - With the volume head and the snapshot included, the `actual size` remains 8.25 Gi.

{{< figure src="/img/screenshots/volumes-and-nodes/volume-size-illustration-fig4.png" >}}

6. Delete data#1 from the mount point.
    - The data#1 filesystem level removal info is stored in current volume head file. For snapshot#1, data#1 is still retained as the historical data.
    - The `actual size` is still 8.25 Gi.

7. Write 8 Gi data (data#2) in the volume mount, then take one more snapshot (snapshot#2). See Figure 5 of the illustration.
    - Now the `actual size` is 16.3 Gi, which is greater than the volume nominal `size`.

>     - There is overlapping between 2 snapshots in block level (See the 2 snapshots in Figure 6), since the data#1 is marked as removed in snapshot#2 then the filesystem reuses the space.

Reviewer: From a filesystem's perspective, the overlapping part (between the 2 snapshots) is considered as the blocks that have to be reused or overwritten. But in terms of Longhorn, these blocks are actually fresh ones held in another snapshot/volume head. See the 2 snapshots in Figure 6.

    > The volume head holds the latest data of the volume only, while each snapshot may store historical data as well as active data, which consumes at most size space. Therefore, the volume `actual size`, which is the size sum of the volume head and all snapshots, is possibly bigger than the size specified by users.
    >
    > Even if users will not take snapshots for volumes, there are operations like rebuilding, expansion, or backing up that would lead to system (hidden) snapshot creation. As a result, volume `actual size` being larger than size is unavoidable under some use cases.

{{< figure src="/img/screenshots/volumes-and-nodes/volume-size-illustration-fig5.png" >}}

8. Delete snapshot#1 and wait for snapshot purge complete. See Figure 7 of the illustration.
    - Here Longhorn actually coalesces the snapshot#1 with the snapshot#2.
    - For the overlapping part during the coalesce, the newer data (data#2) will be retained in the blocks. Then some historical data is removed and the volume gets shrunk (from 16.3 Gi to 11.4 Gi in the example).

{{< figure src="/img/screenshots/volumes-and-nodes/volume-size-illustration-fig6.png" >}}

> Here we summarize the issues we encountered in the example:

Reviewer 2: ```suggestion
Here we summarize the important things related to disk space usage we have in the example:
```

> 1. Unused blocks not releasing issue

Reviewer 2: ```suggestion
1. Unused blocks are not released
```

>     Longhorn does not support TRIM/UNMAP operations. Hence deleting files from filesystems will not lead to volume actual size decreasing/shrinking.

Reviewer 2: Add an extra new line for formatting, otherwise, this will just immediately follow up with the above line.

```suggestion

    Longhorn does not support TRIM/UNMAP operations. Hence deleting files from filesystems will not lead to volume actual size decreasing/shrinking.
```

> 2. Allocated but unused blocks not reusing issue

Reviewer 2: ```suggestion
2. Allocated blocks but unused are not reused
```

>     Deleting then writing new files would lead to the actual size keeps increasing. Since the filesystem may not reuse the recently freed blocks from recently deleted files. Thus, allocating an appropriate nominal size for a volume that holds heavy writing tasks according to the IO pattern would make disk space usage more efficient.

Reviewer 2: ```suggestion

    Deleting then writing new files would lead to the actual size keeps increasing. Since the filesystem may not reuse the recently freed blocks from recently deleted files. Thus, allocating an appropriate nominal size for a volume that holds heavy writing tasks according to the IO pattern would make disk space usage more efficient.
```

3. By deleting snapshots, the overlapping part of the used blocks might be eliminated regardless of whether the blocks are recently released blocks by the filesystem or still contain historical data.
## Space Configuration Suggestions for Volumes

1. Reserve enough free space in disks as buffers in case of the actual size of existing volumes keep growing up.
    - A quick estimation for maximum space consumption estimation of a volume is
  
        ```
        (N + 1 + 1) x head/snapshot average actual size
        ```

>         where N is the number of snapshots retained, and the 1st 1 is for the system snapshot, and 2nd 1 is for the tmp space may required by snapshot deletion.

Reviewer 2: ```suggestion
        where N is the number of snapshots retained, the 1st 1 is for the system snapshot, and the 2nd 1 is for the temporary space that may be required by snapshot deletion.
```

>     - If there are heavy writing tasks for volumes, head/snapshot average actual size would be volume the nominal size. In this case, it's better to set `Storage Over Provisioning Percentage` to be smaller than 100% to avoid disk space exhaustion.

Reviewer 2: ```suggestion
    - If there are heavy writing tasks for volumes, the head/snapshot average actual size would be volume the nominal size. In this case, it's better to set `Storage Over Provisioning Percentage` to be smaller than 100% to avoid disk space exhaustion.
```
